### PR TITLE
ECOM-5421 Add program types to search endpoint

### PR DESCRIPTION
**Pull request by MatthewPiatetsky**
Show program_types in search endpoint so that we will know for the search cards if a course is a MicroMaster

@edx/ecommerce

### CODE REVIEW

File: course_discovery/apps/course_metadata/tests/factories.py [hunk 1]
```diff
@@ -124,6 +124,7 @@ class CourseRunFactory(factory.DjangoModelFactory):
     max_effort = FuzzyInteger(10, 20)
     pacing_type = FuzzyChoice([name for name, __ in CourseRun.PACING_CHOICES])
     slug = FuzzyText()
+    program_types = FuzzyText()
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/tests/factories.py]: This won't do anything.

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -361,6 +361,10 @@ class CourseRun(TimeStampedModel):
     objects = CourseRunQuerySet.as_manager()
 
     @property
+    def program_types(self):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: This is not necessary.

File: course_discovery/apps/course_metadata/search_indexes.py [hunk 3]
```diff
@@ -134,6 +135,9 @@ def prepare_transcript_languages(self, obj):
     def prepare_marketing_url(self, obj):
         return obj.marketing_url
 
+    def prepare_program_types(self, obj):
+        return [p.type.name for p in obj.programs.all()]
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/search_indexes.py]: Avoid unnecessary abbreviations.

Comment (clintonb) [inline course_discovery/apps/course_metadata/search_indexes.py]: Use the property on the model.

File: course_discovery/apps/course_metadata/models.py [hunk 4]
```diff
@@ -361,6 +361,10 @@ class CourseRun(TimeStampedModel):
...
+        return [p.type.name for p in self.programs.all()]
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Avoid unnecessary abbreviations. Also, add a test.

**Comment by schenedx** [comment]
👍

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 5]
```diff
@@ -147,6 +147,11 @@ def test_marketing_url_with_empty_marketing_slug(self, slug):
         course_run = CourseRunFactory(slug=slug)
         self.assertIsNone(course_run.marketing_url)
 
+    def test_program_types(self):
+        """ Verify the property retrieves program types correctly based on programs. """
+        expected = [program.type.name for program in self.course_run.programs.all()]
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/tests/test_models.py]: This is not a good test. Be explicit about the programs you create. Be explicit about the assertion. As-is, `self.course_run` has no programs.

``` python
program = ProgramFactory(courses=[self.course_run.course])
other_program = ProgramFactory(courses=[self.course_run.course])

self.assertEqual(self.course_run.program_types, [program.type.name, other_program.type.name])
```

**Comment by rlucioni** [comment]
👍

File: course_discovery/apps/api/tests/test_serializers.py [hunk 6]
```diff
@@ -610,6 +610,7 @@ def test_data(self):
             'availability': course_run.availability,
             'published': course_run.status == CourseRun.Status.Published,
             'partner': course_run.course.partner.short_code,
+            'program_types': [program.type.name for program in course_run.programs.all()],
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: Associate the course run with some programs to properly test this.

File: course_discovery/apps/course_metadata/search_indexes.py [hunk 7]
```diff
@@ -111,7 +111,8 @@ class CourseRunIndex(BaseCourseIndex, indexes.Indexable):
     seat_types = indexes.MultiValueField(model_attr='seat_types', null=True, faceted=True)
     type = indexes.CharField(model_attr='type', null=True, faceted=True)
     image_url = indexes.CharField(model_attr='card_image_url', null=True)
-    partner = indexes.CharField(null=True, faceted=True)
+    partner = indexes.CharField(model_attr='course__partner__short_code', null=True, faceted=True)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/search_indexes.py]: Does this actually work? If so, remove `prepare_partner`.

File: course_discovery/apps/course_metadata/models.py [hunk 8]
```diff
@@ -306,6 +307,18 @@ def search(cls, query):
         return cls.objects.filter(pk__in=ids)
 
 
+    def save(self, *args, **kwargs):
+        self.index_update(*args, **kwargs)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Call this after save.

File: course_discovery/apps/course_metadata/models.py [hunk 9]
```diff
@@ -306,6 +307,18 @@ def search(cls, query):
...
+        super(Course, self).save(*args, **kwargs)
+
+    def index_update(self, *args, **kwargs):
+        if self.id:
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Remove.

File: course_discovery/apps/course_metadata/models.py [hunk 10]
```diff
@@ -306,6 +307,18 @@ def search(cls, query):
...
+            course_runs = CourseRun.objects.filter(course__id__in=[self.id])
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: self.course_runs

File: course_discovery/apps/course_metadata/models.py [hunk 11]
```diff
@@ -706,6 +719,19 @@ def staff(self):
         return set(staff)
 
 
+    def save(self, *args, **kwargs):
+        self.index_update(*args, **kwargs)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Call after save.

File: course_discovery/apps/course_metadata/models.py (hunk 8 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Also, wrap in an exception handler, logging errors; but not preventing saving.

File: course_discovery/apps/course_metadata/models.py [hunk 12]
```diff
@@ -706,6 +719,19 @@ def staff(self):
...
+        super(Program, self).save(*args, **kwargs)
+
+    def index_update(self, *args, **kwargs):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Remove the unused arguments.

File: course_discovery/apps/course_metadata/models.py [hunk 13]
```diff
@@ -706,6 +719,19 @@ def staff(self):
...
+        if self.id:
+            courses = Course.objects.filter(programs__id__in=[self.id])
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: self.courses

File: course_discovery/apps/course_metadata/search_indexes.py [hunk 14]
```diff
@@ -111,7 +111,7 @@ class CourseRunIndex(BaseCourseIndex, indexes.Indexable):
     seat_types = indexes.MultiValueField(model_attr='seat_types', null=True, faceted=True)
     type = indexes.CharField(model_attr='type', null=True, faceted=True)
     image_url = indexes.CharField(model_attr='card_image_url', null=True)
-    partner = indexes.CharField(model_attr='course__partner__short_code', null=True, faceted=True)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/search_indexes.py]: Why was this changed?

File: course_discovery/apps/course_metadata/models.py [hunk 15]
```diff
@@ -306,6 +307,19 @@ def search(cls, query):
         return cls.objects.filter(pk__in=ids)
 
 
+    def save(self, *args, **kwargs):
+        super(Course, self).save(*args, **kwargs)
+        try:
+            self.index_update()
+        except Exception as e:
+            logger.debug(e)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: `logger.debug` is useless in production. Use `logger.exception` and include a helpful error message.

File: course_discovery/apps/course_metadata/models.py [hunk 16]
```diff
@@ -306,6 +307,19 @@ def search(cls, query):
...
+
+    def index_update(self):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Rename this to represent the action being performed: `reindex_course_runs`. Do the same for the other methods.

File: course_discovery/apps/course_metadata/models.py (hunk 15 above)
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Do the same for the other methods.

File: course_discovery/apps/course_metadata/models.py [hunk 17]
```diff
@@ -306,6 +307,19 @@ def search(cls, query):
...
+        for course_run in self.course_runs.all():
+            index = connections['default'].get_unified_index().get_index(CourseRun)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: Move this outside of the loop.

File: course_discovery/apps/course_metadata/models.py [hunk 18]
```diff
@@ -306,6 +307,19 @@ def search(cls, query):
         return cls.objects.filter(pk__in=ids)
 
 
+    def save(self, *args, **kwargs):
+        super(Course, self).save(*args, **kwargs)
+        try:
+            self.reindex_course_runs()
+        except Exception as e:
+            logger.exception("Exception trying to reindex course runs: {}".format(e))
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py]: No need to include the exception. `logger.exception` does that for you.

Try: `An error occurred while attempting to reindex the course runs of Course [{key}].`

Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/models.py]: If you mean to use format to include the course key in the message, how would I access that if the exception is caught outside the reindex method in the save method?

Comment (clintonb) [reply course_discovery/apps/course_metadata/models.py]: `self.key`

**Comment by MatthewPiatetsky** [comment]
@clintonb tests pass now

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 19]
```diff
@@ -47,6 +48,10 @@ def test_search(self):
         actual = list(Course.search(query).order_by('key'))
         self.assertEqual(actual, courses)
 
+    def test_course_run_update_caught_exception(self):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/tests/test_models.py:51]: Add a docstring. It is not immediately obvious what this test does.

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 20]
```diff
@@ -409,6 +421,10 @@ def test_save_and_no_marketing_site(self):
         self.assertEqual(len(responses.calls), 0)
         toggle_switch('publish_program_to_marketing_site', False)
 
+    def test_course_update_caught_exception(self):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/tests/test_models.py:425]: Add a docstring.

File: course_discovery/apps/course_metadata/models.py [hunk 21]
```diff
@@ -12,6 +12,7 @@
 from django_extensions.db.models import TimeStampedModel
 from djchoices import DjangoChoices, ChoiceItem
 from haystack.query import SearchQuerySet
+from haystack import connections
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:14]: Alphabetize.

**Comment by clintonb** [comment]
👍 (once comments addressed)

**Comment by MatthewPiatetsky** [comment]
think comments are addressed, will merge now to unblock my other ticket, but if there are further comments I can address in another PR
